Adds cudn docs (#85165)
openshift/openshift-docs: modules/nw-cudn-cr.adoc

//
// *networking/multiple_networks/understanding-user-defined-networks.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-cudn-cr_{context}"]
= Creating a ClusterUserDefinedNetwork custom resource

The following procedure creates a cluster-wide user-defined network. Based upon your use case, create your request using either the `cluster-layer-two-udn.yaml` example for a `Layer2` topology type or the `cluster-layer-three-udn.yaml` example for a `Layer3` topology type.

//We won't have these pieces till GA in 4.18.
//[NOTE]
//====
//If any cluster default networked pods exist before the user-defined network is created, any further pods created in this namespace will return an error message: `What_is_this`.
//====

.Procedure

. Create a request for either a `Layer2` or `Layer3` topology type cluster-wide user-defined network:

.. Create a YAML file, such as `cluster-layer-two-udn.yaml`, to define your request for a `Layer2` topology as in the following example:
+
[source, yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: <cudn_name> # <1>
spec:
  namespaceSelector: # <2>
    matchLabels: { "<example_namespace_one>": "<example_namespace_two>" }

>   network:

Reviewer: Should we add text for the network field saying it describe the network configuration, similar to UDN CRD?

Author: Got it.

    topology: Layer2 # <3>
    layer2: <4>
      role: Primary # <5>
      subnets:
        - "2001:db8::/64"
        - "10.100.0.0/16" # <6>
----

> <1> Name of your `ClusterUserDefinedNetwork` resource. This should not be `default` or duplicate of any resource created by the Cluster Network Operator (CNO).

Reviewer: Asking to understand, the `ClusterUserDefinedNetwork` is not managed by CNO, why do we have this restriction?
I dont recall it being validated or prevented by any controller.

Author: Copied from UDN docs. I wasn't sure which callouts (eg <1>, <2>) from those docs would apply to CUDN, and which wouldn't. I have updated.

<6> For `Layer2` topology types the following specifies config details for the `subnet` field:
+
* The subnets field is optional.
* The subnets field is of type `string` and accepts standard CIDR formats for both IPv4 and IPv6.
* The subnets field accepts one or two items. For two items, they must be of a different family. For example, subnets values of `10.100.0.0/16` and `2001:db8::/64`.
* `Layer2` subnets can be omitted. If omitted, users must configure IP addresses for the pods. As a consequence, port security only prevents MAC spoofing.
* The `Layer2` `subnets` field is mandatory when the `ipamLifecycle` field is specified.
+
.. Create a YAML file, such as `cluster-layer-three-udn.yaml`, to define your request for a `Layer3` topology as in the following example:
+
[source, yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: <cudn_name> # <1>
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name <2>
      operator: In <3>
      values: ["<my_value>"] <4>
  network:
    topology: Layer3 # <5>
    layer3: # <6>
      role: Primary # <7>
      subnets: # <8>

>         - cidr: 10.100.0.0/16

Author: Added

<6> This field specifies the topology configuration. Valid values are `layer2` or `layer3`.
<7> Specifies a `Primary` or `Secondary` role. `Primary` is the only `role` specification supported in {product-version}.
<8> For `Layer3` topology types the following specifies config details for the `subnet` field:
+
* The `subnets` field is mandatory.
* The type for the `subnets` field is `cidr` and `hostSubnet`:
** `cidr` is the cluster subnet and accepts a string value.
** `hostSubnet` specifies the nodes subnet prefix that the cluster subnet is split to.
** For IPv6, only a `/64` length is supported for `hostSubnet`.
+
. Apply your request by running the following command:
+
[source,terminal]
----
$ oc apply -f <my_layer_two_udn.yaml>
----
+
Where `<my_layer_two_udn.yaml>` is the name of your `Layer2` or `Layer3` configuration file.

. Verify that your request is successful by running the following command:
+
[source, terminal]
----

> $ oc get clusteruserdefinednetwork <cudn_name> -o yaml

Reviewer: We had some bugs opened, where `oc` is used and the CRD manifest has bad fields names, for example: 
- `mt` instead of `mtu` [[1]](https://issues.redhat.com/browse/OCPBUGS-38706)
- `hostsubnet` instead of `hostSubnet` [[2]](https://issues.redhat.com/browse/OCPBUGS-38901)
It result in unexpected results due to the fact `oc` does not perform schema validation by default [3] - it silently removing fields with bad names and create the object.

Should we specify in the --validate=true in this example? 
Is so, I suppose we should update `UserDefiendNetwork` examples as well.

[1] https://issues.redhat.com/browse/OCPBUGS-38706
[2] https://issues.redhat.com/browse/OCPBUGS-38901
[3] From OC help:
```
 --validate='ignore':
        Must be one of: strict (or true), warn, ignore (or false).              "true" or "strict" will use a schema to validate
        the input and fail the request if invalid. It will perform server side validation if ServerSideFieldValidation
        is enabled on the api-server, but will fall back to less reliable client-side validation if not.                "warn" will
        warn about unknown or duplicate fields without blocking the request if server-side field validation is enabled
        on the API server, and behave as "ignore" otherwise.            "false" or "ignore" will not perform any schema
        validation, silently dropping any unknown or duplicate fields.
```
kubectl set `validation=true` by default.

Author: Added --validate=true to this command.

----
+
Where `<cudn_name>` is the name you created of your cluster-wide user-defined network.
+
.Example output
[source,terminal]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"k8s.ovn.org/v1","kind":"ClusterUserDefinedNetwork","metadata":{"annotations":{},"name":"my-cudn"},"spec":{"namespaceSelector":{"matchExpressions":[{"key":"custom.network.selector","operator":"In","values":["example-namespace-1","example-namespace-2","example-namespace-3"]}]},"network":{"layer3":{"role":"Primary","subnets":[{"cidr":"10.100.0.0/16"}]},"topology":"Layer3"}}}

>   creationTimestamp: "2024-11-19T16:46:34Z"

Reviewer: nit: I would ommit this part, I think its noisy and if you create the manifest using `oc create` this annotation is not added.

Author: Removed lines 114 to 116.

  finalizers:
  - k8s.ovn.org/user-defined-network-protection
  generation: 1
  name: my-cudn
  resourceVersion: "47985"
  uid: 16ee0fcf-74d1-4826-a6b7-25c737c1a634
spec:
  namespaceSelector:
    matchExpressions:
    - key: custom.network.selector
      operator: In
      values:
      - example-namespace-1
      - example-namespace-2
      - example-namespace-3
  network:
    layer3:
      role: Primary
      subnets:
      - cidr: 10.100.0.0/16
    topology: Layer3

> status:

Reviewer: @tssurya @npinaeva should we update docs with the new NetworkAllocated condition? (on a follow-up PR)?

Reviewer 2: updating the docs is good because that status conditional has already merged so it should be part of this PR

[IMPORTANT]
====
The `ClusterUserDefinedNetork` CRD is intended for use by cluster administrators and should not be used by non-administrators. If used incorrectly, it might result in security issues with your deployment, cause disruptions, or break the cluster network. 
====
//We won't have these pieces till GA in 4.18.
//[NOTE]
//====
//If any cluster default networked pods exist before the user-defined network is created, any further pods created in this namespace will return an error message: `What_is_this`.
//====

.Procedure

. Create a request for either a `Layer2` or `Layer3` topology type cluster-wide user-defined network:

.. Create a YAML file, such as `cluster-layer-two-udn.yaml`, to define your request for a `Layer2` topology as in the following example:
+
[source, yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: <cudn_name> # <1>
spec:
  namespaceSelector: # <2>
    matchLabels: # <3>
    - "<example_namespace_one>":"" # <4>
    - "<example_namespace_two>":"" # <4>
  network: # <5>
    topology: Layer2 # <6>
    layer2: # <7>
      role: Primary # <8>
      subnets:
        - "2001:db8::/64"
        - "10.100.0.0/16" # <9>
----

> <1> Name of your `ClusterUserDefinedNetwork` resource. This should not be `default` or a duplicate of any resource in your cluster.

Reviewer 2: right me neither unsure why there should be restrictions on names

> <1> Name of your `ClusterUserDefinedNetwork` resource. This should not be `default` or a duplicate of any resource in your cluster.

Author: Updated to just <1> Name of your `ClusterUserDefinedNetwork` resource.

+
* The subnets field is optional.
* The subnets field is of type `string` and accepts standard CIDR formats for both IPv4 and IPv6.
* The subnets field accepts one or two items. For two items, they must be of a different family. For example, subnets values of `10.100.0.0/16` and `2001:db8::/64`.
* `Layer2` subnets can be omitted. If omitted, users must configure IP addresses for the pods. As a consequence, port security only prevents MAC spoofing.
* The `Layer2` `subnets` field is mandatory when the `ipamLifecycle` field is specified.
+
.. Create a YAML file, such as `cluster-layer-three-udn.yaml`, to define your request for a `Layer3` topology as in the following example:
+
[source, yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: <cudn_name> # <1>
spec:
  namespaceSelector: # <2>
    matchExpressions: # <3>
    - key: kubernetes.io/metadata.name # <4>
      operator: In # <5>
      values: ["<example_namespace_one>, <example_namespace_two>"] # <6>
  network: # <7>
    topology: Layer3 # <8>
    layer3: # <9>
      role: Primary # <10>
      subnets: # <11>
        - cidr: 10.100.0.0/16
          hostSubnet: 64
----

> <1> Name of your `UserDefinedNetwork` resource. This should not be `default` or duplicate any global namespaces created by the Cluster Network Operator (CNO).

Reviewer: > Name of your `UserDefinedNetwork` resource.

ClusterUserDefienedNetwork*

> This should not be `default` 

Same as mentioned above, I am not aware of such restrictions on metadata.name.

> or duplicate any global namespaces created by the Cluster Network Operator (CNO).

I dont undestand what it means or any limitations on the naming related to CNO.

@tssurya do you aware of such restrictions?

Reviewer 2: nope no idea @ormergi don't know any restrictions

Author: Updated to just <1> Name of your `ClusterUserDefinedNetwork` resource.

//module included in the following assembly:
//
// *networking/multiple_networks/primary_networks/about-user-defined-networks.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-cudn-cr_{context}"]
= Creating a ClusterUserDefinedNetwork custom resource

> The following procedure creates a `ClusterUserDefinedResource` custom resource definition (CRD). Based upon your use case, create your request using either the `cluster-layer-two-udn.yaml` example for a `Layer2` topology type or the `cluster-layer-three-udn.yaml` example for a `Layer3` topology type.

Reviewer 3: ```suggestion
The following procedure creates a `ClusterUserDefinedResource` custom resource (CR). Based upon your use case, create your request using either the `cluster-layer-two-udn.yaml` example for a `Layer2` topology type or the `cluster-layer-three-udn.yaml` example for a `Layer3` topology type.
```

It definitely isn't a CRD. These are part of the API extension, and the feature loads the CRD, which is an extension API that provides the ability to create CRs of the type described by the CRD.

So basically we'll always be talking about CRs, and never CRDs.

> The following procedure creates a `ClusterUserDefinedResource` custom resource definition (CRD). Based upon your use case, create your request using either the `cluster-layer-two-udn.yaml` example for a `Layer2` topology type or the `cluster-layer-three-udn.yaml` example for a `Layer3` topology type.

Reviewer 4: don't we mean to say we create a `ClusterUserDefinedNetwork` ?

Author: Fixed

> The following procedure creates a `ClusterUserDefinedResource` custom resource definition (CRD). Based upon your use case, create your request using either the `cluster-layer-two-udn.yaml` example for a `Layer2` topology type or the `cluster-layer-three-udn.yaml` example for a `Layer3` topology type.

Reviewer 4: would it make sense to add a note / disclaimer indicating that for virt use cases the only valid topology is layer2 ? 

quite sure this is mentioned in the virt section of the docs though.

Author: Just to be safe, I added "* {VirtProductName} only supports the `Layer2` topology."

[IMPORTANT]
====

> The `ClusterUserDefinedNetork` CRD is intended for use by cluster administrators and should not be used by non-administrators. If used incorrectly, it might result in security issues with your deployment, cause disruptions, or break the cluster network. 

Reviewer 4: ```suggestion
The `ClusterUserDefinedNetwork` CRD is intended for use by cluster administrators and should not be used by non-administrators. If used incorrectly, it might result in security issues with your deployment, cause disruptions, or break the cluster network. 
```

Author: fixed.

====

.Prerequisites

* You have logged in as a user with `cluster-admin` privileges.

.Procedure

. Optional: For a `ClusterUserDefinedNetwork` CR that uses a primary network, create a namespace with the `k8s.ovn.org/primary-user-defined-network` label by entering the following command:
+
[source,yaml]
----
$ cat << EOF | oc apply -f -
apiVersion: v1
kind: Namespace
metadata:
  name: <cudn_namespace_name>      
  labels:
    k8s.ovn.org/primary-user-defined-network: ""
EOF
----

. Create a request for either a `Layer2` or `Layer3` topology type cluster-wide user-defined network:

.. Create a YAML file, such as `cluster-layer-two-udn.yaml`, to define your request for a `Layer2` topology as in the following example:
+
[source, yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: <cudn_name> # <1>
spec:
  namespaceSelector: # <2>
    matchLabels: # <3>
    - "<example_namespace_one>":"" # <4>
    - "<example_namespace_two>":"" # <4>
  network: # <5>
    topology: Layer2 # <6>

>     layer2: # <7>

Reviewer 4: would we want to expose here the `ipam.lifecycle` attribute ? 

I think it's only useful for virt, so maybe it's OK for us to omit it here, and only refer to it in the virt section of the docs.

Author: Left omitted.

      role: Primary # <8>
      subnets:
        - "2001:db8::/64"
        - "10.100.0.0/16" # <9>
----

> <1> Name of your `ClusterUserDefinedNetwork` resource.

Reviewer 3: ```suggestion
<1> Name of your `ClusterUserDefinedNetwork` custom resource.
```

<2> A label query over the set of namespaces that the cluster UDN applies to. Uses the standard Kubernetes `MatchLabel` selector. Must not point to `default` or `openshift-*` namespaces.

> <3> Uses the `matchLabels` selector type, where terms are evaluated with an _*AND*_ relationship. 

Reviewer 3: Should _and_ be in italics? Or maybe `AND`?

<4> Because the `matchLabels` selector type is used, provisions namespaces matching both `<example_namespace_one>` _and_ `<example_namespace_two>`.
<5> Describes the network configuration.
<6> The `topology` field describes the network configuration; accepted values are `Layer2` and `Layer3`. Specifying a `Layer2` topology type creates one logical switch that is shared by all nodes.
<7> This field specifies the topology configuration. It can be `layer2` or `layer3`.
<8> Specifies `Primary` or `Secondary`. `Primary` is the only `role` specification supported in {product-version}.
<9> For `Layer2` topology types the following specifies config details for the `subnet` field:
+
* The subnets field is optional.
* The subnets field is of type `string` and accepts standard CIDR formats for both IPv4 and IPv6.
* The subnets field accepts one or two items. For two items, they must be of a different family. For example, subnets values of `10.100.0.0/16` and `2001:db8::/64`.

> * `Layer2` subnets can be omitted. If omitted, users must configure IP addresses for the pods. As a consequence, port security only prevents MAC spoofing.

Reviewer 4: ```suggestion
* `Layer2` subnets can be omitted. If omitted, users must configure static IP addresses for the pods. As a consequence, port security only prevents MAC spoofing.
```

Author: Added a link to https://docs.openshift.com/container-platform/4.17/networking/multiple_networks/secondary_networks/creating-secondary-nwt-ovnk.html#configuring-pods-static-ip_configuring-additional-network-ovnk

> * The `Layer2` `subnets` field is mandatory when the `ipamLifecycle` field is specified.

Reviewer 4: so .... I'm torn.

So far, we've hid this attribute. IMHO we either present it before, or ignore it altogether. I'd vote for omitting it.

Then, the API was updated - this is now `ipam.lifecycle`.

Author: I have a separate PR for that here: https://github.com/openshift/openshift-docs/pull/88231, however I will remove it for this PR.

+
.. Create a YAML file, such as `cluster-layer-three-udn.yaml`, to define your request for a `Layer3` topology as in the following example:
+
[source, yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: <cudn_name> # <1>
spec:
  namespaceSelector: # <2>
    matchExpressions: # <3>
    - key: kubernetes.io/metadata.name # <4>
      operator: In # <5>
      values: ["<example_namespace_one>, <example_namespace_two>"] # <6>
  network: # <7>
    topology: Layer3 # <8>
    layer3: # <9>
      role: Primary # <10>
      subnets: # <11>
        - cidr: 10.100.0.0/16
          hostSubnet: 64
----

> <1> Name of your `ClusterUserDefinedNetwork` resource.

Reviewer 3: ```suggestion
<1> Name of your `ClusterUserDefinedNetwork` custom resource.
```

<2> A label query over the set of namespaces that the cluster UDN applies to. Uses the standard Kubernetes `MatchLabel` selector. Must not point to `default` or `openshift-*` namespaces.
<3> Uses the `matchExpressions` selector type, where terms are evaluated with an _*OR*_ relationship. 
<4> Specifies the label key to match.
<5> Specifies the operator. Valid values include: `In`, `NotIn`, `Exists`, and `DoesNotExist`.

> <6> Because the `matchExpressions` type is use, provisions namespaces matching either `<example_namespace_one>` _or_ `<example_namespace_two>`.

Reviewer 3: ```suggestion
<6> Because the `matchExpressions` type is used, provision namespaces matching either `<example_namespace_one>` _or_ `<example_namespace_two>`.
```

Are we using italics for `or`?

<7> Describes the network configuration.
<8> The `topology` field describes the network configuration; accepted values are `Layer2` and `Layer3`. Specifying a `Layer3` topology type creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets.
<9> This field specifies the topology configuration. Valid values are `layer2` or `layer3`.
<10> Specifies a `Primary` or `Secondary` role. `Primary` is the only `role` specification supported in {product-version}.
<11> For `Layer3` topology types the following specifies config details for the `subnet` field:
+
* The `subnets` field is mandatory.
* The type for the `subnets` field is `cidr` and `hostSubnet`:
** `cidr` is the cluster subnet and accepts a string value.
** `hostSubnet` specifies the nodes subnet prefix that the cluster subnet is split to.
** For IPv6, only a `/64` length is supported for `hostSubnet`.
+
. Apply your request by running the following command:
+
[source,terminal]
----

> $ oc create --validate=true -f <example_cluster_udn.yaml>

Reviewer 3: ```suggestion
$ oc create --validate=true -f <example_cluster_udn>.yaml
```

----
+

> Where `<example_cluster_udn.yaml>` is the name of your `Layer2` or `Layer3` configuration file.

Reviewer 3: ```suggestion
Where `<example_cluster_udn>.yaml` is the name of your `Layer2` or `Layer3` configuration file.
```

. Verify that your request is successful by running the following command:
+

> [source, terminal]

Reviewer 3: ```suggestion
[source,terminal]
```

----

> $ oc get clusteruserdefinednetwork <cudn_name> --validate=true -o yaml

Reviewer 4: what's the point of `--validate` while GETing the resource ?... 

I don't think it even exists:
```
oc get --validate
error: unknown flag: --validate
See 'oc get --help' for usage.
```

At least on client version `4.17.14`

Author: I removed it, but if I remember correctly, I couldn't get the CUdn resource without --validate=true. It was weird. at any rate, removed it.

----
+
Where `<cudn_name>` is the name you created of your cluster-wide user-defined network.
+
.Example output
[source,yaml]
----
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  creationTimestamp: "2024-12-05T15:53:00Z"
  finalizers:
  - k8s.ovn.org/user-defined-network-protection
  generation: 1
  name: my-cudn
  resourceVersion: "47985"
  uid: 16ee0fcf-74d1-4826-a6b7-25c737c1a634
spec:
  namespaceSelector:
    matchExpressions:
    - key: custom.network.selector
      operator: In
      values:
      - example-namespace-1
      - example-namespace-2
      - example-namespace-3
  network:
    layer3:
      role: Primary
      subnets:
      - cidr: 10.100.0.0/16
    topology: Layer3
status:
  conditions:
  - lastTransitionTime: "2024-11-19T16:46:34Z"
    message: 'NetworkAttachmentDefinition has been created in following namespaces:
      [example-namespace-1, example-namespace-2, example-namespace-3]'
    reason: NetworkAttachmentDefinitionReady
    status: "True"

>     type: NetworkReady

Reviewer 5: @stevsmit it should be `type: NetworkCreated`

Author: @Meina-rh fixed. Thanks!
